File: SDKs/defold/examples/basic_example.lua
--- IntelliVerseX Defold SDK — complete usage example.
---
--- Demonstrates:
---   1. SDK initialisation
---   2. Device authentication (with session restore)
---   3. Fetching and updating the player profile
---   4. Wallet operations (fetch / grant currency)
---   5. Leaderboard (submit score / fetch records)
---   6. Cloud storage (write / read)
---
--- Usage:
---   Attach this as a script component to a game object in your Defold project.
---   Make sure the IntelliVerseX module is in your project at intelliversex/ivx.lua
---   and the Nakama Defold client is installed as a dependency.

local ivx = require "intelliversex.ivx"

-- Configuration — adjust for your environment
local SERVER_HOST = "127.0.0.1"
local SERVER_PORT = 7350
local SERVER_KEY  = "defaultkey"
local USE_SSL     = false
local DEBUG_LOGS  = true

local LEADERBOARD_ID = "weekly_high_scores"


-- =========================================================================
-- Lifecycle
-- =========================================================================

function init(self)
    math.randomseed(os.time())
    _register_callbacks()
    _configure_sdk()
end

function final(self)
    ivx.disconnect_socket()
    print("[Example] Cleaned up")
end


-- =========================================================================
-- 1. SDK Initialisation
-- =========================================================================

function _configure_sdk()
    ivx.configure({
        host       = SERVER_HOST,
        port       = SERVER_PORT,
        server_key = SERVER_KEY,
        use_ssl    = USE_SSL,
        debug      = DEBUG_LOGS,
    })

    print("[Example] SDK initialized: " .. tostring(ivx.is_initialized()))
    _attempt_auth()
end


-- =========================================================================
-- 2. Authentication
-- =========================================================================

function _attempt_auth()
    if ivx.restore_session() then
        print("[Example] Session restored — user: " .. ivx.get_user_id())
        _post_auth_flow()
    else
        print("[Example] No saved session — authenticating with device ID…")
        ivx.authenticate_device()
    end
end


-- =========================================================================
-- 3. Callbacks
-- =========================================================================

function _register_callbacks()
    ivx.on("auth_success", function(session)
        print("[Example] Auth success! User ID: " .. ivx.get_user_id())
        print("[Example] Username: " .. ivx.get_username())
        _post_auth_flow()
    end)

    ivx.on("auth_error", function(message)
        print("[Example] Auth error: " .. tostring(message))
    end)

    ivx.on("profile", function(profile)
        print("[Example] [signal] Profile loaded for " .. tostring(profile.user_id))
    end)

    ivx.on("wallet", function(wallet)
        print("[Example] [signal] Wallet updated")
    end)

    ivx.on("error", function(message)
        print("[Example] ERROR: " .. tostring(message))
    end)
end


-- =========================================================================
-- 4. Post-auth workflow
-- =========================================================================

function _post_auth_flow()
    _demo_profile()
end


-- =========================================================================
-- 5. Profile
-- =========================================================================

function _demo_profile()
    print("\n--- Profile ---")

    ivx.fetch_profile(function(profile)
        if not profile then
            print("  (could not fetch profile)")
            _demo_wallet()
            return
        end

        print("  Display name : " .. tostring(profile.display_name))
        print("  Avatar URL   : " .. tostring(profile.avatar_url))
        print("  Language      : " .. tostring(profile.lang_tag))

        ivx.update_profile("DefoldPlayer", nil, "en")
        print("  Profile update requested")

        _demo_wallet()
    end)
end


-- =========================================================================
-- 6. Wallet
-- =========================================================================

function _demo_wallet()
    print("\n--- Wallet ---")

    ivx.fetch_wallet(function(wallet)
        print("  Current wallet:")
        if wallet then
            for k, v in pairs(wallet) do
                print("    " .. tostring(k) .. " = " .. tostring(v))
            end
        end

        ivx.grant_currency("coins", 100, function(result)
            print("  Granted 100 coins — result:")
            if result then
                for k, v in pairs(result) do
                    print("    " .. tostring(k) .. " = " .. tostring(v))
                end
            end

            _demo_leaderboard()
        end)
    end)
end


-- =========================================================================
-- 7. Leaderboard
-- =========================================================================

function _demo_leaderboard()
    print("\n--- Leaderboard ---")

    local score = math.random(500, 5000)
    ivx.submit_score(LEADERBOARD_ID, score, function(success)
        print("  Score " .. tostring(score) .. " submitted: " .. tostring(success))

        ivx.fetch_leaderboard(LEADERBOARD_ID, 10, function(records)
            if not records then
                print("  (no records returned)")
                _demo_storage()
                return
            end

            print("  Top " .. tostring(#records) .. " records:")
            for _, r in ipairs(records) do
                print(string.format("    #%s  %s — %d pts",
                    tostring(r.rank),
                    tostring(r.username or "???"),
                    r.score or 0
                ))
            end

            _demo_storage()
        end)
    end)
end


-- =========================================================================
-- 8. Cloud Storage
-- =========================================================================

function _demo_storage()
    print("\n--- Storage ---")

    local save_data = {
        level = 5,
        xp = 2350,
        inventory = { "sword", "shield", "potion" },
        last_save = os.date("%Y-%m-%dT%H:%M:%S"),
    }

    ivx.write_storage("player_saves", "slot_1", save_data, function(success)
        print("  Write OK: " .. tostring(success))

        ivx.read_storage("player_saves", "slot_1", function(loaded)
            if not loaded then
                print("  (nothing loaded)")
            else
                print("  Loaded save — level=" .. tostring(loaded.level) .. "  xp=" .. tostring(loaded.xp))
                if loaded.inventory then
                    print("  Inventory: " .. table.concat(loaded.inventory, ", "))
                end
            end

            print("\n--- All demos complete ---")
        end)
    end)
end

File: SDKs/java/examples/BasicExample.java
package com.intelliversex.sdk.examples;

import com.intelliversex.sdk.core.IVXConfig;
import com.intelliversex.sdk.core.IVXManager;
import com.intelliversex.sdk.core.IVXProfile;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Complete example showing init, auth, profile, wallet, leaderboard, storage, and RPC.
 *
 * <p>Run with a local Nakama server:
 * <pre>{@code
 * javac -cp "lib/*" BasicExample.java
 * java  -cp "lib/*:." com.intelliversex.sdk.examples.BasicExample
 * }</pre>
 */
public class BasicExample {

    public static void main(String[] args) {
        System.out.println("=== IntelliVerseX Java SDK Example ===");
        System.out.println("SDK version: " + IVXManager.SDK_VERSION + "\n");

        // ── 1. Configure ──────────────────────────────────────────────
        IVXConfig config = IVXConfig.builder()
                .nakamaHost("127.0.0.1")
                .nakamaPort(7350)
                .nakamaServerKey("defaultkey")
                .useSSL(false)
                .enableDebugLogs(true)
                .build();

        System.out.println("Config: " + config);

        // ── 2. Initialize ─────────────────────────────────────────────
        IVXManager mgr = IVXManager.getInstance();
        mgr.initialize(config);
        System.out.println("Initialized: " + mgr.isInitialized());

        // ── 3. Restore session or authenticate ────────────────────────
        if (!mgr.restoreSession()) {
            System.out.println("No cached session — authenticating with device ID...");
            mgr.authenticateDevice(null);
        } else {
            System.out.println("Session restored for user: " + mgr.getUserId());
        }

        if (!mgr.hasValidSession()) {
            System.err.println("Authentication failed — exiting.");
            return;
        }

        System.out.println("Authenticated — userId: " + mgr.getUserId()
                + ", username: " + mgr.getUsername());

        // ── 4. Fetch profile ──────────────────────────────────────────
        System.out.println("\n--- Fetching profile ---");
        IVXProfile profile = mgr.fetchProfile();
        if (profile != null) {
            System.out.println("  " + profile);
        } else {
            System.out.println("  (could not fetch profile)");
        }

        // ── 5. Fetch wallet ──────────────────────────────────────────
        System.out.println("\n--- Fetching wallet ---");
        String wallet = mgr.fetchWallet();
        System.out.println("  Wallet JSON: " + wallet);

        // ── 6. Grant currency ─────────────────────────────────────────
        System.out.println("\n--- Granting 100 coins ---");
        String grantResult = mgr.grantCurrency("coins", 100);
        System.out.println("  Grant response: " + grantResult);

        // ── 7. Submit leaderboard score ───────────────────────────────
        System.out.println("\n--- Submitting score ---");
        mgr.submitScore("global_leaderboard", 4200);
        System.out.println("  Score submitted!");

        // ── 8. Fetch leaderboard ──────────────────────────────────────
        System.out.println("\n--- Top 10 leaderboard ---");
        List<Map<String, Object>> records = mgr.fetchLeaderboard("global_leaderboard", 10);
        if (records.isEmpty()) {
            System.out.println("  (no records)");
        } else {
            for (Map<String, Object> r : records) {
                System.out.printf("  #%s  %s  score=%s%n",
                        r.get("rank"), r.get("username"), r.get("score"));
            }
        }

        // ── 9. Storage round-trip ─────────────────────────────────────
        System.out.println("\n--- Storage write/read ---");
        mgr.writeStorage("game_data", "player_prefs", "{\"volume\":0.8,\"difficulty\":\"hard\"}");
        String stored = mgr.readStorage("game_data", "player_prefs");
        System.out.println("  Read back: " + stored);

        // ── 10. Custom RPC ────────────────────────────────────────────
        System.out.println("\n--- Custom RPC ---");
        String rpcResult = mgr.callRpc("my_custom_rpc", "{\"action\":\"ping\"}");
        System.out.println("  RPC response: " + rpcResult);

        // ── 11. Async example ─────────────────────────────────────────
        System.out.println("\n--- Async profile fetch ---");
        CompletableFuture<IVXProfile> asyncProfile = mgr.fetchProfileAsync();
        asyncProfile
                .thenAccept(p -> System.out.println("  Async profile: " + p))
                .exceptionally(ex -> {
                    System.err.println("  Async profile error: " + ex.getMessage());
                    return null;
                })
                .join();

        // ── 12. Cleanup ──────────────────────────────────────────────
        System.out.println("\n--- Clearing session ---");
        mgr.clearSession();
        System.out.println("  Session valid: " + mgr.hasValidSession());

        System.out.println("\n=== Example complete ===");
    }
}
